dialog/fileitem.go

```diff
@@ -35,44 +41,67 @@ func (i *fileDialogItem) TappedSecondary(_ *fyne.PointEvent) {
 }
 
 func (i *fileDialogItem) CreateRenderer() fyne.WidgetRenderer {
-	img := canvas.NewImageFromResource(i.icon)
+	var img fyne.CanvasObject
+	if i.icon == nil {
+		mimeSub := i.mimeSubType
+		if len(mimeSub) > 5 {
+			mimeSub = mimeSub[:5]
+		}
+		img = NewFileIcon(i.mimeType, mimeSub)
+	} else {
+		img = canvas.NewImageFromResource(i.icon)
+	}
 	text := widget.NewLabelWithStyle(i.name, fyne.TextAlignCenter, fyne.TextStyle{})
 	text.Wrapping = fyne.TextTruncate
-	extText := canvas.NewText(i.ext, theme.BackgroundColor())
-	extText.Alignment = fyne.TextAlignCenter
-	extText.TextSize = theme.TextSize()
+
 	return &fileItemRenderer{item: i,
-		img: img, text: text, ext: extText, objects: []fyne.CanvasObject{img, text, extText}}
+		img: img, text: text, objects: []fyne.CanvasObject{img, text}}
 }
 
 func (i *fileDialogItem) isDirectory() bool {
 	return i.icon == theme.FolderIcon() || i.icon == theme.FolderOpenIcon()
 }
 
-func fileParts(path string) (name, ext string) {
+func fileParts(path string) (name, ext, mimeType, mimeSubType string) {
 	name = filepath.Base(path)
 	ext = filepath.Ext(name[1:])
 	name = name[:len(name)-len(ext)]
 
-	if len(ext) > 1 {
-		ext = ext[1:]
+	mimeTypeFull := mime.TypeByExtension(ext)
+	if mimeTypeFull == "" {
+		mimeTypeFull = "text/plain"
+		file, err := os.Open(path)
+		if err == nil {
+			defer file.Close()
+			scanner := bufio.NewScanner(file)
+			if scanner.Scan() && !utf8.Valid(scanner.Bytes()) {
+				mimeTypeFull = "application/octet-stream"
+			}
+		}
 	}
+
+	mimeTypeSplit := strings.Split(mimeTypeFull, "/")
+	mimeType = mimeTypeSplit[0]
+	mimeSubType = mimeTypeSplit[1]
+
 	return
 }
 
 func (f *fileDialog) newFileItem(icon fyne.Resource, path string) *fileDialogItem {
-	name, ext := fileParts(path)
```

Comment:
Perhaps moving this code into fileDialogItem is tidier? As far as I can see they are only used in the CreateRenderer of the child object, so don't have to be passed in and managed as state of the item.
I might be wrong, just looks that way from this diff